PhpAmqpLib/Wire/IO/StreamIO.php

```diff
@@ -132,40 +174,60 @@ public function reconnect()
     }
 
     /**
-     * @param $n
+     * @param $len
      * @throws \PhpAmqpLib\Exception\AMQPIOException
      * @return mixed|string
      */
-    public function read($n)
+    public function read($len)
     {
-        $res = '';
         $read = 0;
+        $data = '';
 
-        while ($read < $n && !feof($this->sock) && (false !== ($buf = fread($this->sock, $n - $read)))) {
+        while ($read < $len) {
             $this->check_heartbeat();
 
-            if ($buf === '') {
+            if (!is_resource($this->sock) || feof($this->sock)) {
+                throw new AMQPRuntimeException('Broken pipe or closed connection');
+            }
+
+            set_error_handler(array($this, 'error_handler'));
+            $buffer = fread($this->sock, ($len - $read));
+            restore_error_handler();
+
+            if ($buffer === false) {
+                throw new AMQPRuntimeException('Error receiving data');
+            }
+
+            if ($buffer === '') {
                 if ($this->canDispatchPcntlSignal) {
-                    pcntl_signal_dispatch();
+                    // prevent cpu from being consumed while waiting
+                    if ($this->canSelectNull) {
+                        $this->select(null, null);
+                        pcntl_signal_dispatch();
+                    } else {
+                        usleep(100000);
+                        pcntl_signal_dispatch();
+                    }
                 }
                 continue;
             }
 
-            $read += mb_strlen($buf, 'ASCII');
-            $res .= $buf;
-
-            $this->last_read = microtime(true);
+            $read += mb_strlen($buffer, 'ASCII');
+            $data .= $buffer;
         }
 
-        if (mb_strlen($res, 'ASCII') != $n) {
-            throw new AMQPIOException(sprintf(
-                'Error reading data. Received %s instead of expected %s bytes',
-                mb_strlen($res, 'ASCII'),
-                $n
-            ));
+        if (mb_strlen($data, 'ASCII') !== $len) {
+            throw new AMQPRuntimeException(
```

Comment:
Any reason why you changed the Exception type? If you have to change it, I think you need to update the PHP Doc section of this function